File: src/spec/resources/testprojects/pytest/petstore_client/default_api_client.py
from typing import Any, Dict, Optional
import gzip
import uuid
import zlib

import urllib3

from petstore_client.api_client import ApiClient
from petstore_client.api_response import ApiResponse
from petstore_client.configuration import Configuration

try:
    import brotli as _brotli  # type: ignore[import-untyped]
except ImportError:
    _brotli = None

try:
    import zstandard as _zstandard
except ImportError:
    _zstandard = None  # type: ignore[assignment]


def _supported_encodings() -> str:
    encodings = ['gzip', 'deflate']
    if _brotli is not None:
        encodings.append('br')
    if _zstandard is not None:
        encodings.append('zstd')
    return ', '.join(encodings)


class DefaultApiClient:
    """Default implementation of ApiClient using urllib3."""

    def __init__(self, config: Optional[Configuration] = None, pool_manager: Optional[Any] = None) -> None:
        if pool_manager is not None:
            self._pool_manager = pool_manager
        elif config is not None:
            kwargs = {}
            if not config.verify_ssl:
                kwargs['cert_reqs'] = 'CERT_NONE'
            elif config.ssl_ca_cert:
                kwargs['ca_certs'] = config.ssl_ca_cert
                kwargs['cert_reqs'] = 'CERT_REQUIRED'
            if config.proxy:
                self._pool_manager = urllib3.ProxyManager(config.proxy, **kwargs)
            else:
                self._pool_manager = urllib3.PoolManager(**kwargs)
        else:
            self._pool_manager = urllib3.PoolManager()

    def send_request(self, method: str, url: str, headers: Dict[str, str], body: Any = None) -> ApiResponse:
        """Send an HTTP request and return the response.

        :param method: HTTP method (GET, POST, PUT, DELETE, etc.)
        :param url: Fully qualified URL
        :param headers: HTTP headers
        :param body: Request body (serialized JSON string, bytes, dict for multipart, or None)
        :return: ApiResponse containing status code, body, and headers
        """
        headers.setdefault('Accept-Encoding', _supported_encodings())

        if isinstance(body, dict):
            boundary = str(uuid.uuid4())
            headers['Content-Type'] = f'multipart/form-data; boundary={boundary}'
            encoded_body = self._build_multipart_body(body, boundary)
        elif isinstance(body, bytes):
            encoded_body = body
        elif isinstance(body, str):
            encoded_body = body.encode('utf-8')
        else:
            encoded_body = None

        response = self._pool_manager.request(
            method, url, headers=headers, body=encoded_body, preload_content=False, decode_content=False
        )

        raw_data = response.read()
        content_encoding = (response.headers.get('content-encoding') or '').lower()
        decompressed = self._decompress_body(raw_data, content_encoding)
        response_body = decompressed.decode('utf-8') if decompressed else ''
        response_headers = dict(response.headers) if response.headers else {}

        return ApiResponse(status_code=response.status, body=response_body, headers=response_headers)

    def _build_multipart_body(self, form_parts: Dict[str, Any], boundary: str) -> bytes:
        """Build a multipart/form-data body from a dict of form parts.

        :param form_parts: dict mapping field names to values
        :param boundary: the multipart boundary string
        :return: encoded multipart body as bytes
        """
        parts: list[bytes] = []
        for name, value in form_parts.items():
            if isinstance(value, list):
                for item in value:
                    parts.append(self._multipart_part(name, item, boundary))
            else:
                parts.append(self._multipart_part(name, value, boundary))
        parts.append(f'--{boundary}--\r\n'.encode('utf-8'))
        return b''.join(parts)

    @staticmethod
    def _multipart_part(name: str, value: Any, boundary: str) -> bytes:
        """Build a single MIME part for multipart encoding.

        :param name: the field name
        :param value: the field value (str, bytes, file-like, or model object)
        :param boundary: the multipart boundary string
        :return: encoded MIME part as bytes
        """
        if hasattr(value, 'read'):
            raw_data = value.read()
            file_data: bytes = raw_data.encode('utf-8') if isinstance(raw_data, str) else raw_data
            header = (
                f'--{boundary}\r\n'
                f'Content-Disposition: form-data; name="{name}"; filename="{name}"\r\n'
                f'Content-Type: application/octet-stream\r\n\r\n'
            )
            return header.encode('utf-8') + file_data + b'\r\n'
        elif isinstance(value, bytes):
            header = (
                f'--{boundary}\r\n'
                f'Content-Disposition: form-data; name="{name}"; filename="{name}"\r\n'
                f'Content-Type: application/octet-stream\r\n\r\n'
            )
            return header.encode('utf-8') + value + b'\r\n'
        elif hasattr(value, 'model_dump_json'):
            json_str: str = value.model_dump_json(by_alias=True, exclude_none=True)
            header = (
                f'--{boundary}\r\n'
                f'Content-Disposition: form-data; name="{name}"\r\n'
                f'Content-Type: application/json\r\n\r\n'
            )
            return header.encode('utf-8') + json_str.encode('utf-8') + b'\r\n'
        else:
            part = f'--{boundary}\r\nContent-Disposition: form-data; name="{name}"\r\n\r\n{value}\r\n'
            return part.encode('utf-8')

    @staticmethod
    def _decompress_body(data: bytes, encoding: str) -> bytes:
        """Decompress response body based on Content-Encoding header.

        Supports gzip and deflate natively. Brotli and zstd are supported
        when their respective packages (brotli, zstandard) are installed.

        :param data: raw response bytes
        :param encoding: Content-Encoding header value
        :return: decompressed bytes
        """
        if not data:
            return data
        if encoding in ('gzip', 'x-gzip'):
            return gzip.decompress(data)
        if encoding == 'deflate':
            return zlib.decompress(data)
        if encoding == 'br' and _brotli is not None:
            br_result: bytes = _brotli.decompress(data)
            return br_result
        if encoding == 'zstd' and _zstandard is not None:
            zstd_result: bytes = _zstandard.ZstdDecompressor().decompress(data, max_output_size=len(data) * 16)
            return zstd_result
        return data

File: src/spec/resources/testprojects/nodetest/src/default-api-client.ts
import type { ApiClient } from './api-client.js';
import type { ApiResponse } from './api-response.js';
import type { Configuration } from './configuration.js';
import * as crypto from 'node:crypto';
import * as https from 'node:https';
import * as http from 'node:http';
import * as fs from 'node:fs';
import * as zlib from 'node:zlib';
import { promisify } from 'node:util';

/**
 * Default implementation of ApiClient using the Fetch API.
 *
 * When a Configuration with proxy or TLS settings is provided,
 * falls back to Node.js http/https modules for full control.
 */
export class DefaultApiClient implements ApiClient {
  private readonly config?: Configuration;
  private readonly agent?: https.Agent;

  constructor(config?: Configuration) {
    this.config = config;
    if (config && (config.sslCaCert != null || !config.verifySsl)) {
      this.agent = new https.Agent({
        ca: config.sslCaCert ? fs.readFileSync(config.sslCaCert) : undefined,
        rejectUnauthorized: config.verifySsl
      });
    }
  }

  async sendRequest(
    method: string,
    url: string,
    headers: Record<string, string>,
    body: string | Buffer | null
  ): Promise<ApiResponse> {
    if (body instanceof Blob) {
      body = Buffer.from(await body.arrayBuffer());
    }

    headers['Accept-Encoding'] ??= DefaultApiClient.supportedEncodings();

    if (body != null && typeof body === 'object' && !Buffer.isBuffer(body)) {
      const boundary = crypto.randomUUID();
      headers['Content-Type'] = `multipart/form-data; boundary=${boundary}`;
      body = await this.buildMultipartBody(body as Record<string, unknown>, boundary);
    }

    if (this.config?.proxy || this.agent) {
      return this.sendWithNodeHttp(method, url, headers, body);
    }
    return this.sendWithFetch(method, url, headers, body);
  }

  private async sendWithFetch(
    method: string,
    url: string,
    headers: Record<string, string>,
    body: string | Buffer | null
  ): Promise<ApiResponse> {
    const response = await fetch(url, {
      method,
      headers,
      body: body ?? undefined
    });

    const responseBody = await response.text();
    const responseHeaders: Record<string, string> = {};
    response.headers.forEach((value, key) => {
      responseHeaders[key] = value;
    });

    return {
      statusCode: response.status,
      body: responseBody,
      headers: responseHeaders
    };
  }

  private sendWithNodeHttp(
    method: string,
    url: string,
    headers: Record<string, string>,
    body: string | Buffer | null
  ): Promise<ApiResponse> {
    return new Promise((resolve, reject) => {
      const parsed = new URL(url);
      const isHttps = parsed.protocol === 'https:';

      let requestOptions: http.RequestOptions;

      if (this.config?.proxy) {
        const proxyUrl = new URL(this.config.proxy);
        if (isHttps) {
          const connectReq = http.request({
            host: proxyUrl.hostname,
            port: Number(proxyUrl.port) || 3128,
            method: 'CONNECT',
            path: `${parsed.hostname}:${parsed.port || 443}`
          });

          connectReq.on('connect', (_res, socket) => {
            const tlsOptions: https.RequestOptions = {
              host: parsed.hostname,
              port: Number(parsed.port) || 443,
              path: parsed.pathname + parsed.search,
              method,
              headers,
              createConnection: () => socket,
              agent: this.agent ?? new https.Agent({ rejectUnauthorized: this.config?.verifySsl ?? true })
            };
            const req = https.request(tlsOptions, (res) => this.collectResponse(res, resolve));
            req.on('error', reject);
            if (body) req.write(body);
            req.end();
          });

          connectReq.on('error', reject);
          connectReq.end();
          return;
        } else {
          requestOptions = {
            host: proxyUrl.hostname,
            port: Number(proxyUrl.port) || 3128,
            path: url,
            method,
            headers: { ...headers, Host: parsed.host }
          };
        }
      } else {
        requestOptions = {
          hostname: parsed.hostname,
          port: parsed.port ? Number(parsed.port) : undefined,
          path: parsed.pathname + parsed.search,
          method,
          headers,
          agent: isHttps ? this.agent : undefined
        };
      }

      const transport = this.config?.proxy || !isHttps ? http : https;
      const req = transport.request(requestOptions, (res) => this.collectResponse(res, resolve));
      req.on('error', reject);
      if (body) req.write(body);
      req.end();
    });
  }

  private collectResponse(res: http.IncomingMessage, resolve: (value: ApiResponse) => void): void {
    const chunks: Buffer[] = [];
    res.on('data', (chunk: Buffer) => chunks.push(chunk));
    res.on('end', () => {
      const raw = Buffer.concat(chunks);
      const encoding = (res.headers['content-encoding'] ?? '').toLowerCase();
      this.decompressBody(raw, encoding).then((decompressed) => {
        const responseBody = decompressed.toString('utf-8');
        const responseHeaders: Record<string, string> = {};
        for (const [key, value] of Object.entries(res.headers)) {
          if (typeof value === 'string') {
            responseHeaders[key] = value;
          } else if (Array.isArray(value)) {
            responseHeaders[key] = value.join(', ');
          }
        }
        resolve({
          statusCode: res.statusCode ?? 0,
          body: responseBody,
          headers: responseHeaders
        });
      });
    });
  }

  static supportedEncodings(): string {
    const encodings = ['gzip', 'deflate', 'br'];
    if (typeof zlib.zstdDecompress === 'function') {
      encodings.push('zstd');
    }
    return encodings.join(', ');
  }

  private async decompressBody(data: Buffer, encoding: string): Promise<Buffer> {
    if (data.length === 0) return data;
    switch (encoding) {
      case 'gzip':
      case 'x-gzip':
        return promisify(zlib.gunzip)(data);
      case 'deflate':
        return promisify(zlib.inflate)(data);
      case 'br':
        return promisify(zlib.brotliDecompress)(data);
      case 'zstd':
        if (typeof zlib.zstdDecompress === 'function') {
          return promisify(zlib.zstdDecompress)(data);
        }
        return data;
      default:
        return data;
    }
  }

  private async buildMultipartBody(formParts: Record<string, unknown>, boundary: string): Promise<Buffer> {
    const parts: Buffer[] = [];
    for (const [name, value] of Object.entries(formParts)) {
      if (Array.isArray(value)) {
        for (const item of value) {
          parts.push(await this.multipartPart(name, item, boundary));
        }
      } else {
        parts.push(await this.multipartPart(name, value, boundary));
      }
    }
    parts.push(Buffer.from(`--${boundary}--\r\n`, 'utf-8'));
    return Buffer.concat(parts);
  }

  private async multipartPart(name: string, value: unknown, boundary: string): Promise<Buffer> {
    if (Buffer.isBuffer(value)) {
      const header = `--${boundary}\r\nContent-Disposition: form-data; name="${name}"; filename="${name}"\r\nContent-Type: application/octet-stream\r\n\r\n`;
      return Buffer.concat([Buffer.from(header, 'utf-8'), value, Buffer.from('\r\n', 'utf-8')]);
    }
    if (value instanceof Blob) {
      const buf = Buffer.from(await value.arrayBuffer());
      const header = `--${boundary}\r\nContent-Disposition: form-data; name="${name}"; filename="${name}"\r\nContent-Type: application/octet-stream\r\n\r\n`;
      return Buffer.concat([Buffer.from(header, 'utf-8'), buf, Buffer.from('\r\n', 'utf-8')]);
    }
    if (typeof value === 'object' && value !== null) {
      const json = JSON.stringify(value);
      return Buffer.from(
        `--${boundary}\r\nContent-Disposition: form-data; name="${name}"\r\nContent-Type: application/json\r\n\r\n${json}\r\n`,
        'utf-8'
      );
    }
    return Buffer.from(
      `--${boundary}\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${String(value)}\r\n`,
      'utf-8'
    );
  }
}
